libtenzir/include/tenzir/package.hpp

```diff
@@ -0,0 +1,190 @@
+
+  auto to_record() const -> record;
+
+  static auto parse(const view<record>& data) -> caf::expected<package_input>;
+
+  friend auto inspect(auto& f, package_input& x) -> bool {
+    return f.object(x)
+      .pretty_name("package_input")
+      .fields(f.field("name", x.name), f.field("description", x.description),
+              f.field("type", x.type), f.field("default", x.default_));
+  }
+};
+
+struct package_pipeline final {
+  std::string name = {};
+  std::string description = {};
+  std::string definition = {}; // required
+  bool disabled = false;
+  std::optional<duration> retry_on_error = {};
+
+  auto to_record() const -> record;
+
+  static auto
+  parse(const view<record>& data) -> caf::expected<package_pipeline>;
+
+  friend auto inspect(auto& f, package_pipeline& x) -> bool {
+    return f.object(x)
+      .pretty_name("package_pipeline")
+      .fields(f.field("name", x.name), f.field("description", x.description),
+              f.field("definition", x.definition),
+              f.field("disabled", x.disabled),
+              f.field("retry_on_error", x.retry_on_error));
+  }
+};
+
+struct package_context final {
+  std::string type = {};
+  std::string description = {};
+  tenzir::context::parameter_map arguments = {};
+  bool disabled = false;
+
+  auto to_record() const -> record;
+
+  static auto parse(const view<record>& data) -> caf::expected<package_context>;
+
+  friend auto inspect(auto& f, package_context& x) -> bool {
+    return f.object(x)
+      .pretty_name("package_context")
+      .fields(f.field("type", x.type), f.field("description", x.description),
+              f.field("arguments", x.arguments),
+              f.field("disabled", x.disabled));
+  }
+};
+
+struct package_snippet final {
+  std::string name = {};
+  std::string description = {};
+  std::string definition = {}; // required
+
+  auto to_record() const -> record;
+
+  static auto parse(const view<record>& data) -> caf::expected<package_snippet>;
+
+  friend auto inspect(auto& f, package_snippet& x) -> bool {
+    return f.object(x)
+      .pretty_name("package_snippet")
+      .fields(f.field("name", x.name), f.field("description", x.description),
+              f.field("definition", x.definition));
+  }
+};
+
+struct package_inputs_map
+  : public detail::flat_map<std::string, package_input> {
+  using super = detail::flat_map<std::string, package_input>;
+  using super::super;
+};
+
+struct package_pipelines_map
+  : public detail::flat_map<std::string, package_pipeline> {
+  using super = detail::flat_map<std::string, package_pipeline>;
+  using super::super;
+};
+
+struct package_contexts_map
+  : public detail::flat_map<std::string, package_context> {
+  using super = detail::flat_map<std::string, package_context>;
+  using super::super;
+};
+
+using package_snippets_list = std::vector<package_snippet>;
+
+struct package final {
+  std::string id = {};   // required
+  std::string name = {}; // required
```

Comment:
If the other fields are not required, shouldn't they be modeled as `std::optional<std::string>`?

Reply:
From the app's perspective there is a difference, I think, whether there is an author or not. And `show packages | where author != null` does also not work if the author is rendered as an empty string.